internal/tools/aws/database_multitenant.go

```diff
@@ -0,0 +1,637 @@
+				Values: []*string{aws.String(dbClusterID)},
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get RDS cluster id %s", dbClusterID)
+	}
+	if len(dbClusterOutput.DBClusters) != 1 {
+		return nil, fmt.Errorf("expected exactly one RDS cluster for installation id %s (found %d)", d.installationID, len(dbClusterOutput.DBClusters))
+	}
+
+	return dbClusterOutput.DBClusters[0], nil
+}
+
+func (d *RDSMultitenantDatabase) acquireLock(dbClusterID string, store model.InstallationDatabaseStoreInterface) (func(logger log.FieldLogger), error) {
+	locked, err := store.LockDatabaseCluster(dbClusterID, d.installationID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not acquire lock for database cluster %s", dbClusterID)
+	}
+	if !locked {
+		return nil, errors.Errorf("could not acquire lock for database cluster %s", dbClusterID)
+	}
+	unlockFN := func(logger log.FieldLogger) {
+		unlocked, err := store.UnlockDatabaseCluster(dbClusterID, d.installationID, true)
+		if err != nil {
+			logger.WithError(err).Errorf("provisioner datastore failed to release locker for database cluster ID %s", dbClusterID)
+		}
+		if !unlocked {
+			logger.Warnf("database ID %s and locker ID %s is still locked", dbClusterID, d.installationID)
+		}
+	}
+
+	return unlockFN, nil
+}
+
+func (d *RDSMultitenantDatabase) validateDatabaseClusterInstallations(installations model.DatabaseClusterInstallations, databaseCluster *model.DatabaseCluster) error {
+	expectedInstallations, err := databaseCluster.GetInstallations()
+	if err != nil {
+		return errors.Errorf("failed to get installations from database cluster ID %s", databaseCluster.ID)
+	}
+
+	if installations.Size() != expectedInstallations.Size() {
+		return errors.Errorf("supplied %s installations, but the cluster ID %s has %s", installations.Size(), expectedInstallations.Size(), databaseCluster.ID)
+	}
+
+	for _, installation := range installations {
+		if !expectedInstallations.Contains(installation) {
+			return errors.Errorf("unable to find installation ID %s in cluster ID %s", installation, databaseCluster.ID)
+		}
+	}
+
+	return nil
+}
+
+func (d *RDSMultitenantDatabase) connectRDSCluster(schema, endpoint, username, password string) (func(logger log.FieldLogger), error) {
+	// This condition allows injecting our own driver for testing.
+	if d.db == nil {
+		db, err := sql.Open("mysql", RDSMySQLConnString(schema, endpoint, username, password))
+		if err != nil {
+			return nil, errors.Wrap(err, "connecting to RDS DB cluster")
+		}
+
+		d.db = db
+	}
+
+	close := func(logger log.FieldLogger) {
+		err := d.db.Close()
+		if err != nil {
+			logger.WithError(err).Errorf("failed to close database connection to %s", endpoint)
+		}
+	}
+
+	return close, nil
+}
+
+func (d *RDSMultitenantDatabase) createUserIfNotExist(ctx context.Context, username, password string) error {
+	query := fmt.Sprintf("CREATE USER IF NOT EXISTS '%s'@'%s' IDENTIFIED BY '%s'", username, "%", password)
```

Comment:
This type of query construction is vulnerable to SQL injection. Instead of `Sprintf` and `%s`, you need to use `?` for string interpolation in conjunction with the SQL library like this: 

https://github.com/mattermost/mattermost-cloud/blob/master/internal/store/installation.go#L64

Reply:
I know, but we can't use the placeholder for "CREATE", apparently. I tried and it errors. It seems that there is a bug somewhere in  the sql library. I have to find it.

Reply:
I did some investigation on this. When you connect to MySQL you have to add a parameter to allow parameter interpolation and then you can use parameters. The following is equivalent to your `createUserIfNotExists` using `database/sql` and the MySQL driver, except I'm using `Exec` instead of `ExecWithContext` just to simplify my example.

```go
db, err := sql.Open("mysql", "user:my-secret-pw@/mysql?interpolateParams=true")
q := "CREATE USER IF NOT EXISTS ?@? IDENTIFIED BY ?"
result, err := db.Exec(q, "ian", "%", "mypassword")
```

I ran a local MySQL Docker container and was able to execute the above successfully with the Go REPL `gore`.

Reply:
This is the error that I'm seeing when using the placeholder `Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?@? IDENTIFIED BY ?' at line 1`. However, `interpolateParams=true` did the trick. Thanks!

Reply:
Ok, too early. Can you try `result, err := db.Exec("CREATE DATABASE IF NOT EXISTS ? CHARACTER SET ?", databaseName, "utf8mb4")`?

Reply:
@gigawhitlocks It turns out that **this is a bug or at very least a limitation** of the library in place. The issue here is that when using the placeholders the library *always input single quote* which results in an invalid [SQL statement](https://dev.mysql.com/doc/refman/5.7/en/create-database.html).

db.Exec("CREATE DATABASE IF NOT EXISTS ?", "my_database", "utf8mb4"):
```
Invalid:

CREATE DATABASE IF NOT EXISTS 'my_database';
```

```
valid:

CREATE DATABASE IF NOT EXISTS my_database;
```

> I ran a local MySQL Docker container and was able to execute the above successfully with the Go REPL gore

I'm not sure if you tried the other two commands but they result in the same error:
```
 Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''cloud_ubzzobtfginpdq773nw3chr1cw'' at line 1
```

> TL;DR func RDSMySQLConnString in helpers.go should be modified to add &interpolateParams=true to the returned value and then you can change these three queries to use SQL-injection resistant interpolation.

I agree that we should always write code without vulnerabilities. However, injection is normally a concern for applications which take human inputs. In the case of this application, database, user creating and permissions is exclusively being done by the provisioner. Besides  to that, I don't think that (CREATE USER, CREATE DATABASE, and GRANT) are susceptible to [SQL-injection](https://owasp.org/www-community/attacks/SQL_Injection).